### PR TITLE
[native] Add support for mutable System Config

### PR DESCRIPTION
**Pull request by spershin**
Adding support to Native Worker to have mutable System Config.
It is activated by having "mutable-config=true" in config.properties file.
Examples of use:
```
curl "127.0.0.1:7777/v1/operation/systemConfig/setProperty?name=shutdown-onset-sec&value=2"
curl "myhost:7777/v1/operation/systemConfig/setProperty?name=shutdown-onset-sec&value=2"
curl "127.0.0.1:7777/v1/operation/systemConfig/getProperty?name=shutdown-onset-sec"
curl "myhost:7777/v1/operation/systemConfig/getProperty?name=shutdown-onset-sec"
```
Test plan - new tests.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/common/Configs.cpp [hunk 1]
```diff
@@ -29,11 +29,38 @@ ConfigBase::ConfigBase()
...
   filePath_ = filePath;
 }
 
+folly::Optional<std::string> ConfigBase::setValue(
+    const std::string& propertyName,
+    const std::string& value) {
+  auto oldValue = config_->get(propertyName);
+  if (auto* memConfig =
+          dynamic_cast<velox::core::MemConfigMutable*>(config_.get())) {
+    memConfig->setValue(propertyName, value);
```
Comment (tanjialiang) [inline presto-native-execution/presto_cpp/main/common/Configs.cpp:55]: maybe: return oldValue; here
and remove else

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 2]
```diff
@@ -58,8 +65,9 @@ class ConfigBase {
     return config_->get(propertyName);
   }
 
-  const std::unordered_map<std::string, std::string>& values() const {
-    return config_->values();
+  /// Returns copy of the config values map.
+  std::unordered_map<std::string, std::string> values() const {
```
Comment (tanjialiang) [inline presto-native-execution/presto_cpp/main/common/Configs.h:69]: why this change?

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 3]
```diff
@@ -750,6 +753,27 @@ std::string PrestoServer::connectorOperation(
   }
 }
 
+std::string PrestoServer::systemConfigOperation(
+    const ServerOperation& op,
+    proxygen::HTTPMessage* message) {
+  switch (op.action) {
+    case ServerOperation::Action::kSetProperty: {
```
[the remark this answers is not in the document]
Comment (spershin) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:760]: I thought about it.
We can add any time we need one.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 4]
```diff
@@ -750,6 +753,27 @@ std::string PrestoServer::connectorOperation(
...
+      const auto name = message->getQueryParam("name");
+      const auto value = message->getQueryParam("value");
```
[the remark this answers is not in the document]
Comment (spershin) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:762]: What type of validation do you have in mind?
Just being non-empty?
Because validation for specific settings - that is likely not feasible for all props.
Keep in mind that this is only for dev clusters, not for production.

File: presto-native-execution/presto_cpp/main/common/Configs.cpp [hunk 5]
```diff
@@ -29,11 +29,38 @@ ConfigBase::ConfigBase()
     : config_(std::make_unique<velox::core::MemConfig>()) {}
 
 void ConfigBase::initialize(const std::string& filePath) {
-  config_ = std::make_unique<velox::core::MemConfig>(
-      util::readConfig(fs::path(filePath)));
+  // See if we want to create a mutable config.
```
[the remark this answers is not in the document]
Comment (spershin) [reply presto-native-execution/presto_cpp/main/common/Configs.cpp:32]: Yes.

**Comment by tanjialiang** [comment]
Thanks Sergey, some minor comments

File: presto-native-execution/presto_cpp/main/common/Configs.cpp (hunk 1 above)
Comment (spershin) [reply presto-native-execution/presto_cpp/main/common/Configs.cpp:55]: Will all compilers be happy with missing return?
I'd rather leave it.

File: presto-native-execution/presto_cpp/main/common/Configs.h (hunk 2 above)
Comment (spershin) [reply presto-native-execution/presto_cpp/main/common/Configs.h:69]: Because underneath can be a mutable config and we cannot return reference to the map which then can be looped through while someone changing it from another thread.
This method is only called during startup.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:760]: I think set and get should be hand in hand. When experimenting I think it might be good to see before set.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:762]: No like, if the user does not provide "name" param or "value" param. The message should indicate the accurate error instead of letting SystemConfig handle it and throw a confusing message.

[the remark this answers is not in the document]
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:762]: Maybe can also add in test about this scenario.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (spershin) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:760]: Sure, I can add that here.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (spershin) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:762]: How can we add tests with http commands, do we have one already?

File: presto-native-execution/presto_cpp/main/common/Configs.h (hunk 2 above)
Comment (tanjialiang) [reply presto-native-execution/presto_cpp/main/common/Configs.h:69]: But there is no protection either when you are making the copy in valuesCopy() method. Do we want to add a lock if we want it to be safe?

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 6]
```diff
@@ -27,6 +27,13 @@ class ConfigBase {
   /// @param filePath Path to configuration file.
   void initialize(const std::string& filePath);
 
+  /// Adds or replaces value at the given key. Can be used by debugging or
+  /// testing code.
+  /// Returns previous value if there was any.
+  folly::Optional<std::string> setValue(
```
Comment (gggrace14) [inline presto-native-execution/presto_cpp/main/common/Configs.h:33]: Should we add the setValue() to the subclass SystemConfig & NodeConfig, instead of the base class ConfigBase? Similar to the initialize() function. If added to the base class, it means all the existing and new subclasses of ConfigBase will all be mutable if mutable-config is set. However, we haven't implemented the ServerOperation endpoint for NodeConfig yet. Is it your planned design?

File: presto-native-execution/presto_cpp/main/common/Configs.h (hunk 2 above)
Comment (spershin) [reply presto-native-execution/presto_cpp/main/common/Configs.h:69]: The lock is inside config_->valuesCopy();

**Review by gggrace14 — COMMENTED** [review]
@spershin , a few nit-picking comments left.

The main thing I see is, have you checked which system config properties are mutable by this design? I see several places, ex., TaskManager, are using an instance of SystemConfig or systemConfig->values(). My understanding is those won't get muted by calling the endpoint. If it is the case, should we enumerate the list somehow and return some message when the user tries to set those properties? Let me know if my understanding is not correct.

File: presto-native-execution/presto_cpp/main/common/Configs.cpp [hunk 7]
```diff
@@ -29,11 +29,38 @@ ConfigBase::ConfigBase()
...
+
   filePath_ = filePath;
 }
 
+folly::Optional<std::string> ConfigBase::setValue(
+    const std::string& propertyName,
+    const std::string& value) {
+  auto oldValue = config_->get(propertyName);
+  if (auto* memConfig =
+          dynamic_cast<velox::core::MemConfigMutable*>(config_.get())) {
```
[the remark this answers is not in the document]
Comment (spershin) [reply presto-native-execution/presto_cpp/main/common/Configs.cpp:53]: 1. The casts are equivalent.
2. We have unique pointer, so I'm not quite sure how ptr_cast will produce another unique pointer here, unless we use ref type in the cast.
3. We only need pointer, don't care about smart pointer. :)

File: presto-native-execution/presto_cpp/main/ServerOperation.cpp [hunk 8]
```diff
@@ -16,23 +16,31 @@
 
 namespace facebook::presto {
 
-const std::unordered_map<std::string, ServerOperation::Action>
-    ServerOperation::kActionLookup = {
+const folly::F14FastMap<std::string, ServerOperation::Action>
```
[the remark this answers is not in the document]
Comment (spershin) [reply presto-native-execution/presto_cpp/main/ServerOperation.cpp:19]: Just following the established pattern here.

File: presto-native-execution/presto_cpp/main/common/Configs.h (hunk 6 above)
Comment (spershin) [reply presto-native-execution/presto_cpp/main/common/Configs.h:33]: Having such support in the base class simplifies things for any config based on it.

> If added to the base class, it means all the existing and new subclasses of ConfigBase will all be mutable if mutable-config is set. However, we haven't implemented the ServerOperation endpoint for NodeConfig yet. Is it your planned design?

Yes, if we need that,
In fact I plan something like that for the Query Config too, but that stuff is complicated (check the Task, if interested).
So, we'll see how it will pan out.

**Comment by spershin** [comment]
> One thing I see is, have you checked which system config properties are mutable by this change? I see several places, ex., TaskManager, are using systemConfig->values() or initialized by passing individual properties. My understanding is those won't get muted by calling the endpoint. If it is the case, should we enumerate the list somehow and return some message when the user tries to set those properties? Let me know if my understanding is not correct.

The idea is the following:
1. All properties are mutable.
2. The ones that we read only once are also mutable, but change will have no effect.
3. The ones that are used to init permanent events (such as ssd cache) are also mutable, but change will have no effect.
4. We need to get away of caching system properties at startup (like TaskManager does) and move all such code sites (if possible) to read them when needed.

Let me know if that makes sense.

**Comment by gggrace14** [comment]
> > One thing I see is, have you checked which system config properties are mutable by this change? I see several places, ex., TaskManager, are using systemConfig->values() or initialized by passing individual properties. My understanding is those won't get muted by calling the endpoint. If it is the case, should we enumerate the list somehow and return some message when the user tries to set those properties? Let me know if my understanding is not correct.
> 
> The idea is the following:
> 
> 1. All properties are mutable.
> 2. The ones that we read only once are also mutable, but change will have no effect.
> 3. The ones that are used to init permanent events (such as ssd cache) are also mutable, but change will have no effect.
> 4. We need to get away of caching system properties at startup (like TaskManager does) and move all such code sites (if possible) to read them when needed.
> 
> Let me know if that makes sense.
a
4 is the definitely the right way to go. I'm thinking there might not be many properties left if not 2 or 3 at the moment. Anyway, approving.

**Review by gggrace14 — APPROVED** [review]

**Comment by gggrace14** [comment]
Thank you, Sergey @spershin , for implementing this!

**Comment by spershin** [comment]
> Thank you, Sergey @spershin , for implementing this!

@gggrace14 
Thank you for reviewing!
I hope we can wrap it up the next week and have support for modifying all the properties we need.
